### PR TITLE
[WIP] EHN/TST advance TTR

### PR DESCRIPTION
**Pull request by glemaitre**
TODO:

- [ ] More tests
- [ ] Documentation

### CODE REVIEW

**Comment by amueller** [comment]
cool. So you don't want to do a PR to master and finish it off?

File: sklearn/preprocessing/label.py [hunk 1]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+
+    Parameters
+    ----------
+    estimator : object, (default=LinearRegression())
+        Estimator object derived from ``RegressorMixin``.
+
+    transformer : object, (default=None)
+        Estimator object derived from ``TransformerMixin``. Cannot be set at
+        the same time as ``func`` and ``inverse_func``. If ``None`` and
+        ``func`` and ``inverse_func`` as well, the transformer will be an
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: are ``None`` as well

File: sklearn/preprocessing/label.py [hunk 2]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+        identity transformer.
 
-    Parameter
-    ---------
-    estimator : object
-        Estimator object to wrap.
+    func : function, (default=None)
+        Function to apply to ``y`` before passing to ``fit``. Cannot be set at
+        the same time than ``transformer``. If ``None`` and ``transformer`` as
+        well, the function used will be the identity function.
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: is None as well..

File: sklearn/preprocessing/label.py [hunk 3]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
 
-    func : function
-        Function to apply to y before passing to fit.
+    inverse_func : function, (default=None)
+        Function apply to the prediction of the estimator. Cannot be set at
+        the same time than ``transformer``. If ``None`` and ``transformer`` as
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: same

File: sklearn/preprocessing/label.py [hunk 4]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+        Whether to check that ``transform`` followed by ``inverse_transform``
+        or ``func`` followed by ``inverse_func`` lead to the original data.
 
     Attributes
     ----------
     estimator_ : object
         Fitted estimator.
 
+    transformer_ : object
+        Used transformer in ``fit`` and ``predict``.
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: Only present if ``transformer`` is not None? Or always?

File: sklearn/preprocessing/label.py [hunk 5]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
     array([ 2.])
 
-
-    Notes
-    -----
-    It is not checked whether func and inverse_func are actually
-    inverse to each other.
     """
-    def __init__(self, estimator, func, inverse_func):
+    def __init__(self, estimator, transformer=None,
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: Didn't you say LinearRegression is the default estimator? That might be good for now before we got better common tests.

File: sklearn/preprocessing/label.py [hunk 6]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("Both 'transformer' and functions 'func'/"
+                             "'inverse_func' cannot be set at the same time.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        self.transformer_.fit(y)
+        # XXX: is it a necessary test? one might not want an invertible
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: I don't think you need the comment as we have the flag now.

File: sklearn/preprocessing/label.py [hunk 7]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+        # function.
+        if self.check_invertible:
+            n_subsample = min(1000, y.shape[0])
+            subsample_idx = np.random.randint(y.shape[0], size=n_subsample)
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: why not permutation? randint does bootstrap samples (with repetition)

File: sklearn/preprocessing/label.py [hunk 8]
```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
...
+            diff = np.abs((y[subsample_idx] -
+                           self.transformer_.inverse_transform(
+                               self.transformer_.transform(y[subsample_idx]))))
+            if np.sum(diff) > 1e-7:
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly invertible.")
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: hm maybe say the functions are not the inverse of each other? they might be invertible but not be the right functions.

File: sklearn/preprocessing/label.py [hunk 9]
```diff
@@ -138,12 +189,23 @@ def fit(self, X, y, sample_weight=None):
         self : object
             Returns self.
         """
+        # memorize if y should be a multi-output
+        self.y_shape_ = y.shape
+        if len(y.shape) == 1:
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: y.ndim

File: sklearn/preprocessing/label.py [hunk 10]
```diff
@@ -138,12 +189,23 @@ def fit(self, X, y, sample_weight=None):
         self : object
             Returns self.
         """
+        # memorize if y should be a multi-output
+        self.y_shape_ = y.shape
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: hm maybe we shouldn't remember ``y.shape[0]``? Though dummy tries that and then has two flags ``n_outputs_`` and ``outputs_2d_``.

File: sklearn/preprocessing/label.py [hunk 11]
```diff
@@ -154,38 +216,56 @@ def predict(self, X):
...
-            from ..metrics import accuracy_score
-            return accuracy_score(y, self.predict(X),
-                                  sample_weight=sample_weight)
-        elif is_regressor(self.estimator):
-            from ..metrics import r2_score
-            return r2_score(y, self.predict(X), sample_weight=sample_weight,
-                            multioutput='variance_weighted')
-        else:
-            # I'm not sure if this is too many internals for an error message
+        if not is_regressor(self.estimator_):
```
Comment (amueller) [inline sklearn/preprocessing/label.py]: Maybe we should even remove this? I guess it prevents people from using his with a classifier and then being surprised by R2, so on second though leave it in.

File: sklearn/preprocessing/tests/test_label.py [hunk 12]
```diff
@@ -511,3 +517,44 @@ def test_inverse_binarize_multiclass():
...
+                         " are not strictly invertible.", clf.fit, X, y)
+    clf = TransformedTargetRegressor(estimator=LinearRegression(),
+                                     func=np.exp, inverse_func=np.exp,
+                                     check_invertible=False)
+    # the transformer/functions are not checked to be invertible the fitting
+    # should pass
+    clf.fit(X, y)
+
+
+def test_target_transformer_friedman():
```
Comment (amueller) [inline sklearn/preprocessing/tests/test_label.py]: There is no test that the functions are actually applied, right?

Comment (amueller) [inline sklearn/preprocessing/tests/test_label.py]: That, and as you said, you should have one with y.shape = (n_samples, 1) and one with y.shape = (n_samples, 3) or something.

File: sklearn/preprocessing/label.py (hunk 4 above)
Comment (glemaitre) [inline sklearn/preprocessing/label.py]: always since we create a transformer from the functions

**Pull request closed without merging**